### ISSUE TITLE
Bugs on multi party secure computation

### ISSUE THREAD

**Issue by fu1001hao** [OpenMined/PySyft#5663]
## Description
The new pysyft, SyMPC. We can't let MPCTensor divide by a  constant or minus by a constant. In the 2.9.* version, it does not have such problem. Also, when we share the neural network to session, the model.parameters() return empty list because the model contains two sub-neural networks. But it shouldn't be the case. It should return all the parameters.

## How to Reproduce
2. Click on '...'
3. Scroll down to '...'

## System Information
 - OS Version: [e.g. 22]
 - Language Version: [e.g. Python 3.7, Node 10.18.1]
 - Package Manager Version: [e.g. Conda 4.6.1, NPM 6.14.1]
 - Browser (if applicable): [e.g. Google Chrome]
 - Browser Version (if applicable): [e.g. 81.0.4044.138]

**Comment by rasswanth-s** [comment]
@fu1001hao  ,maybe could you post a screenshot of the error , when you do operations of  MPCTensor with a constant,also syft version

**Comment by fu1001hao** [comment]
> @fu1001hao ,maybe could you post a screenshot of the error , when you do operations of MPCTensor with a constant,also syft version

<img width="817" alt="image" src="https://user-images.githubusercontent.com/57201660/122649242-f2de7a00-d0fa-11eb-89b3-ed2fae625d11.png">

The version is "0.5.0rc2.post1.dev25+ge7fa7629f.d20210520"

There are other issues. For example, if a network structure (model) contains two subnetworks (model1, model2), the MPC will return an empty list of parameters, when calling model.parameters(). You have to use model1.parameters()+model2.paramters() to call the parameters. Also, the MPCTensor cannot be divided by constant, it will return None type data.

**Comment by rasswanth-s** [comment]
@fu1001hao I will take a look at it. could you also post the whole code, where you get None value from MPCTensor.

**Comment by fu1001hao** [comment]
> @fu1001hao I will take a look at it. could you also post the whole code, where you get None value from MPCTensor, when doing operations with constant.

<img width="438" alt="image" src="https://user-images.githubusercontent.com/57201660/122649823-cd06a480-d0fd-11eb-8d38-28348dfcc621.png">

You set "batchs" to be any constant number, you get None type. I also attach the whole code.

import syft as sy
import torch
import torch.optim as optim
from sympc.session import Session
from sympc.session import SessionManager
from sympc.module import nn
from sympc.optim import SGD
from sympc.tensor import MPCTensor
import time

start = time.time()
sy.load("sympc")
sy.logger.add(sink="./example.log")

alice = sy.VirtualMachine(name="alice")
bob = sy.VirtualMachine(name="bob")

alice_client = alice.get_client()
bob_client = bob.get_client()

session = Session(parties=[alice_client, bob_client])
session.autograd_active = True
SessionManager.setup_mpc(session)

x, y = torch.load('data/train.pt')
xt, yt = torch.load('data/test.pt')

def one_hot(y):

    tmp = torch.zeros((len(y),10))
    for i in range(len(y)):
        tmp[i, y[i]] = 1
    return tmp

def split(x, y):

    party0, party1 = [], []
    for i in range(10):
        indices = (y==i).nonzero()[:,0]
        l = len(indices) // 2
        party0.append(indices[:l])
        party1.append(indices[l:])
    party0, party1 = torch.cat(party0, 0), torch.cat(party1, 0)
    return (x[party0], y[party0]), (x[party1], y[party1])

party0, party1 = split(x, y)

class Net(sy.Module):

    def __init__(self, torch_ref):
        super(Net, self).__init__(torch_ref = torch_ref)
        self.fc1 = torch_ref.nn.Linear(28*28, 128)
        self.fc2 = torch_ref.nn.Linear(128, 64)
        self.fc3 = torch_ref.nn.Linear(64, 10)

    def forward(self, x):
        x = self.torch_ref.nn.functional.relu(self.fc1(x))
        x = self.torch_ref.nn.functional.relu(self.fc2(x))
        x = self.fc3(x)
        
        return x

def share(party, pct = 0.01):

    x, y = party
    shared = []
    for i in range(10):
        indices = (y==i).nonzero()[:,0]
        l = int(len(indices)*pct)
        shared.append(indices[:l])
    shared = torch.cat(shared, 0)
    return x[shared], y[shared]


model = Net(torch_ref = torch)
private = True

x0, y0 = party0
xs, ys = share(party1)
x0, xs = x0.reshape(-1, 28*28), xs.reshape(-1, 28*28)
print(x0.shape, xs.shape)

ys = one_hot(ys)
y0 = one_hot(y0)
lr = 1e-2
epoch = 500
device = 'cpu'

batch0, batchs = y0.shape[0], ys.shape[0]
x0, y0 = x0.to(device), y0.to(device)
xs, ys = xs.to(device), ys.to(device)

shapexs, shapeys = xs.shape, ys.shape

if private:

    xs = xs.share(session = session)
    ys = ys.share(session = session)

for i in range(epoch):

    print(i)
    for j in range(2):

        if (j==0):

            if i!=0 and private:

                model = model.reconstruct()

            optimizer = optim.SGD(model.parameters(), lr = lr)
            optimizer.zero_grad()
            output = model(x0)
            loss = ((output - y0)**2).mean()
            loss.backward()
            optimizer.step()
            print(loss.item())

        elif (j==1):

            if private:

                model = model.share(session)
                optimizer = SGD(model.parameters(), lr = lr)
            else:

                optimizer = optim.SGD(model.parameters(), lr = lr)
                
            optimizer.zero_grad()
            output = model(xs)
            loss = ((output - ys)**2).sum()/batchs
            loss.backward()
            optimizer.step()
            if private:

                print(loss.reconstruct())

            else:

                print(loss.item())

xt = xt.reshape(-1, 28*28)
xt, yt = xt.to(device), yt.to(device)
if private:

    model = model.reconstruct()
output = model(xt)
pred = output.argmax(dim=1)
total = pred.eq(yt).sum().item()
print(total*1.0/len(pred), total, len(pred))

print(time.time() - start)

**Comment by gmuraru** [comment]
If you take the latest version of SyMPC - from the dev branch - do you have the same problem?
It might be that at the point of releasing there was no support for division?

**Comment by fu1001hao** [comment]
> If you take the latest version of SyMPC - from the dev branch - do you have the same problem?
> It might be that at the point of releasing there was no support for division?

 I downloaded via this link: https://github.com/OpenMined/SyMPC/tree/main/src/sympc. Is this the latest one?

**Comment by rasswanth-s** [comment]
You could install the latest version by `pip install git+https://github.com/OpenMined/SyMPC@dev` , it would install latest one.

**Comment by fu1001hao** [comment]
> You could install the latest version by `pip install git+https://github.com/OpenMined/SyMPC@dev` , it would install latest one.

<img width="762" alt="image" src="https://user-images.githubusercontent.com/57201660/122684666-b97a3d00-d1d4-11eb-85e2-1f9d75858ce7.png">

It doesn't work.

**Comment by rasswanth-s** [comment]
Sorry ,I interchanged in pysyft we `dev` for SyMPC we use `main` , this would work `pip install git+https://github.com/OpenMined/SyMPC@main`

**Comment by fu1001hao** [comment]
> Sorry ,I interchanged in pysyft we `dev` for SyMPC we use `main` , this would work `pip install git+https://github.com/OpenMined/SyMPC@main`

I got a totally different error. Before update I could at least run the code by not using dividing, now I can't even run the code.

<img width="920" alt="image" src="https://user-images.githubusercontent.com/57201660/122811106-12fa6e80-d29e-11eb-939c-00ce0f17a742.png">

I even ran the 'introduction.ipynb' on your SyMPC GitHub. It gave me the same error.

**Comment by rasswanth-s** [comment]
Yeah,I answered the issue for a similar error ,could you follow these issues https://github.com/OpenMined/PySyft/issues/5701 and https://github.com/OpenMined/PySyft/issues/5671

**Comment by fu1001hao** [comment]
> Yeah,I answered the issue for a similar error ,could you follow these issues #5701 and #5671

Not work for me. I run

pip uninstall syft
pip uninstall sympc

then run

pip install -e "git+https://github.com/OpenMined/PySyft@dev#egg=syft&subdirectory=packages/syft"
pip install git+https://github.com/OpenMined/SyMPC@main

And it shows successful install but still when I run the code I got the same error.

**Comment by rasswanth-s** [comment]
@fu1001hao ,Could you do it in a new virtual env?

**Comment by fu1001hao** [comment]
I did, but the problem was not solved.

The error is:
<img width="672" alt="image" src="https://user-images.githubusercontent.com/57201660/122942535-47c10100-d344-11eb-8a04-fd261aa6e041.png">
because I did division in my code:
<img width="563" alt="image" src="https://user-images.githubusercontent.com/57201660/122942648-5e675800-d344-11eb-83b7-6f05c093f985.png">

If I remove the division, it has no error. So overall, even the latest version does not support division.

**Comment by rasswanth-s** [comment]
The latest version , supports division, @gmuraru  is it because of spdz issue with division(2 parties),?

**Comment by gmuraru** [comment]
The division works for the moment with a number of two parties (we are working towards adding support for more)

**Comment by fu1001hao** [comment]
> The division works for the moment with a number of two parties (we are working towards adding support for more)

I uninstalled pysyft and sympc, then reinstalled. It shows the same error. From my test,  only the MPCTensor that are processed by neural network will become NoneType by division (which is weird). If I just simply divide an MPCTensor by a constant, it works well (I have checked that). So I think it is from the sy.Module or MPC version of the network that causes the issue.
